contracts/router/src/contract.rs

```diff
@@ -417,28 +419,35 @@ fn simulate_swap_operations(
 
                 // Deduct tax before querying simulation
                 if let AssetInfo::NativeToken { denom } = offer_asset_info.clone() {
-                    offer_amount =
-                        offer_amount.checked_sub(compute_tax(deps, offer_amount, denom)?)?;
+                    let asset = Asset {
+                        info: AssetInfo::NativeToken { denom },
+                        amount: offer_amount,
+                    };
+
+                    offer_amount = offer_amount.checked_sub(asset.compute_tax(&deps.querier)?)?;
                 }
 
                 let mut res: SimulationResponse =
                     deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
                         contract_addr: pair_info.contract_addr.to_string(),
                         msg: to_binary(&PairQueryMsg::Simulation {
                             offer_asset: Asset {
-                                info: offer_asset_info,
+                                info: offer_asset_info.clone(),
                                 amount: offer_amount,
                             },
                         })?,
                     }))?;
 
                 // Deduct tax after querying simulation
-                if let AssetInfo::NativeToken { denom } = ask_asset_info.clone() {
-                    res.return_amount = res.return_amount.checked_sub(compute_tax(
-                        deps,
-                        res.return_amount,
-                        denom,
-                    )?)?;
+                if let AssetInfo::NativeToken { .. } = ask_asset_info.clone() {
+                    let asset = Asset {
+                        info: offer_asset_info,
+                        amount: offer_amount,
```

Comment:
Why did you switch to offer asset instead of ask here?

Reply:
Fixed. Thanks